R/genius_url.R

```diff
@@ -18,67 +18,92 @@
 #' @importFrom stringr str_detect str_extract
 #' @importFrom readr read_lines
 
-genius_url <- function(url, info = "title") {
- session <- html_session(url)
- # Clean the song lyrics
- lyrics <- gsub(pattern = "<.*?>",
-                replacement = "\n",
-                html_node(session, ".lyrics")) %>%
-    read_lines() %>%
-    na.omit() %>%
-    str_replace_all("’", "'")
+genius_url <- function(url, info = "title")  {
+  # create a new session for scraping lyrics
+  session <- html_session(url)
 
-# Artist
-artist <- html_nodes(session, ".header_with_cover_art-primary_info-primary_artist") %>%
+  # Get Artist name
+  artist <- html_nodes(session, ".header_with_cover_art-primary_info-primary_artist") %>%
     html_text() %>%
     str_replace_all("\n", "") %>%
     str_trim()
 
-# Song title
-song_title <- html_nodes(session, ".header_with_cover_art-primary_info-title") %>%
+  # Get Song title
+  song_title <- html_nodes(session, ".header_with_cover_art-primary_info-title") %>%
     html_text() %>%
     str_replace_all("\n", "") %>%
     str_trim()
 
-# Convert to tibble
-lyrics <- tibble(artist = artist,
-            track_title = song_title,
-            lyric = lyrics)
+  # scrape the lyrics
+  lyrics <- # read the text from the lyrics class
+    html_node(session, ".lyrics") %>%
+    # trim white space
+    html_text(trim = TRUE) %>%
+    # use named vector for cleaning purposes
+    str_replace_all(cleaning()) %>% {
+      # sometimes there is only one line in a song
+      # if vector length == one it will try to read the text as a filepath
+      # add blank text if it that long
+      if (length(.) == 1) {
+        . <- c(.,"")
```

Comment:
This bit is somewhat awkward but is necessary for handling songs that are instrumental. This will keep an NA row for lyric but will maintain the element and element artist field